fabric-events-interaction-v0/src/main/java/net/fabricmc/fabric/api/entity/FakePlayer.java

```diff
@@ -0,0 +1,154 @@
+ *
+ * <p>The easiest way to obtain a fake player is with {@link FakePlayer#get(ServerWorld)} or {@link FakePlayer#get(ServerWorld, GameProfile)}.
+ * It is also possible to create a subclass for more control over the fake player's behavior.
+ *
+ * <p>For good inter-mod compatibility, fake players should have the UUID of their owning (human) player.
+ * They should still have a different name to ensure the {@link GameProfile} is different.
+ * For example:
+ * <pre>{@code
+ * UUID humanPlayerUuid = ...;
+ * String humanPlayerName = ...;
+ * GameProfile fakeProfile = new GameProfile(humanPlayerUuid, "[Block Breaker of " + humanPlayerName + "]");
+ * }</pre>
+ * If a fake player does not belong to a specific player, the {@link #DEFAULT_UUID default UUID} should be used.
+ *
+ * <p>Fake players try to behave like regular {@link ServerPlayerEntity} objects to a reasonable extent.
+ * In some edge cases, or for gameplay considerations, it might be necessary to check whether a {@link ServerPlayerEntity} is a fake player.
+ * This can be done with an {@code instanceof} check: {@code player instanceof FakePlayer}.
+ */
+public class FakePlayer extends ServerPlayerEntity {
+	/**
+	 * Default UUID, for fake players not associated with a specific (human) player.
+	 */
+	public static final UUID DEFAULT_UUID = UUID.fromString("41C82C87-7AfB-4024-BA57-13D2C99CAE77");
+	private static final GameProfile DEFAULT_PROFILE = new GameProfile(DEFAULT_UUID, "[Minecraft]");
+
+	/**
+	 * Retrieves a fake player for the specified world, using the {@link #DEFAULT_UUID default UUID}.
+	 * This is suitable when the fake player is not associated with a specific (human) player.
+	 * Otherwise, the UUID of the owning (human) player should be used (see class javadoc).
+	 *
+	 * <p>Instances are reused for the same world parameter.
+	 *
+	 * <p>Caution should be exerted when storing the returned value,
+	 * as strong references to the fake player will keep the world loaded.
+	 */
+	public static FakePlayer get(ServerWorld world) {
+		return get(world, DEFAULT_PROFILE);
+	}
+
+	/**
+	 * Retrieves a fake player for the specified world and game profile.
+	 * See class javadoc for more information on fake player game profiles.
+	 *
+	 * <p>Instances are reused for the same parameters.
+	 *
+	 * <p>Caution should be exerted when storing the returned value,
+	 * as strong references to the fake player will keep the world loaded.
+	 */
+	public static FakePlayer get(ServerWorld world, GameProfile profile) {
+		Objects.requireNonNull(world, "World may not be null.");
+		Objects.requireNonNull(profile, "Game profile may not be null.");
+
+		return FAKE_PLAYER_MAP.computeIfAbsent(new FakePlayerKey(world, profile), key -> new FakePlayer(key.world, key.profile));
+	}
+
+	private record FakePlayerKey(ServerWorld world, GameProfile profile) { }
+	private static final Map<FakePlayerKey, FakePlayer> FAKE_PLAYER_MAP = new MapMaker().weakValues().makeMap();
```

Comment:
Won't this leak the worlds anyway because the values are what are made weak while `FakePlayerKey`s aren't but they hold a reference to the `ServerWorld`?

It should be cleared at some point.

Reply:
> Or does the value being GC'd somehow trigger the key to be to?

Indeed, that is the whole point. 😉